### DISCUSSION TITLE
How to preprocess html templates in app built on the quasar framework

### DISCUSSION THREAD

**Discussion by ohchi** [quasarframework/quasar#15998] [General - Components / Directives / etc] [answered]
My future web app is a part of the firmware. The firmware is supposed to be built for a lot of devices. For each of them there is a profile with configuration parameters. I need to clean up app's html templates from unnecessary features. The reason i want to do it at build time is the size of firmware. So, i want to replace all `v-if` statements that use `profile` object with constant values.
For example, this html  
```
<div v-if="profile.feature1"></div>
```
is supposed to be replaced with this:
```
<div v-if="false"></div>
```
After this the whole div tag will be removed as dead.
As far as I understand, I can do a similar thing with js by writing a plugin for `rollup`, but with html... no idea.

In previous version of this web app i used `gulp` plugin as a part of html processing pipeline.

Which options do i have if i migrate to quasar?

PS: 
I figured out that i have no need to write `rollup` plugin in case of js. 
According to docs:

> When compiling your website/app, if () branches depending on process.env are evaluated, and if the expression is false, they get stripped out of the file. 

That's a good news. But i still don't know how to access `process.env` in templates and would it be stripped like js.

**Comment by freddy38510** [comment]
Why not defining the value of `profile.feature1` outside templates ?
```js
// profile.js
const profile = {
  feature1: process.env.FEATURE1 // this will be replaced during compilation
};

export { profile }
```

Then in your Vue SFC:
```vue
<template>
<div v-if="profile.feature1">feature1 is supported</div>
</template>

<script setup>
import { profile } from './profile.js';
</script>
```

```js
// quasar.config.js
module.exports = function (ctx) {
  return {
    // ...

    build: {
      env: {
        FEATURE1: process.env.FEATURE1
      }
    }
  }
}
```

Finally you can enable/disable `FEATURE1` before building your app:
```bash
# we set an env variable in terminal
$ FEATURE1=true quasar build
```

**Reply by ohchi** [reply]
@freddy38510 thank you for reply. Would it be tree shaked in this case?

**Reply by freddy38510** [reply]
I don't think so, because the v-if directive is used to conditionally render a block **at runtime**.

As I understand it, the code has to be statically parsable to be "treeshaked". But you can always try.

Another approach would be to use [dynamic components](https://vuejs.org/guide/essentials/component-basics.html#dynamic-components) for each device.

**Reply by ohchi** [reply]
Dynamic components are not the choice in my case, unfortunately. I have a whole lot of small features (usually turning into several fields on a page) which originated from small differences in devices (for instance now we have about 300 profile constants) and customer's requirements as well. Though each separate difference is small, as a rule, but altogether they end up taking considerable space on the flash.

**Comment by pdanpdan** [comment]
If you use render functions and the condition is based on values known at build (env constants) it should be removed

**Reply by ohchi** [reply]
That's the reason i talk about finding the way to access env constants in templates. I haven't found this way by now.

**Comment by ohchi** [comment]
I found another way to preprocess sources. There is `rawDefine` property in `quasar.config.js`. AFAIK it's based on the same `define` `vite`'s feature. It's allows me to use `SOME_FEATURE`  constants in source code directly without `process.env`. It didn't give the answer for templates by itself, but i looked at what came to `define` plugin as input and saw `_ctx.SOME_FEATURE`. So, now i can add `_ctx.` prefixed version of every constant in `rawDefine` to make it replaced by `define` plugin.

The problems that i still have:
1. I don't know how reliable it is. It looks like a workaround.
2. The tag with `v-if="SOME_FEATURE"` is stripped, but if the tag refers to a component and that component isn't used anywhere else it's still here. On other hand, if i replace `v-if="SOME_FEATURE"` to `v-if="false"` then both (the tag of the component and the component itself) are removed.

**Comment by freddy38510** [comment]
Did you try `v-if="SOME_FEATURE || false"` ?

Direct use of [env variables](https://v2.vitejs.dev/guide/env-and-mode.html) seems more appropriate for your use case. Because `import.meta.env.NON_EXISTENT` should be replaced as `undefined`. In this case, there's no need to use a prefix.

**Reply by ohchi** [reply]
I don't have problems with fallback values for my constants anymore. Since i decided to use `constants.d.ts` for declaring all global constants i'm able to generate files with initialization for all ways of using these constants. No more implicit constants. 

For runtime:
For templates i generate `app.config.globalProperties`  initialization in `boot` folder. For js code it's a file with global js constants that is loaded by script tag in `index.html`.

For build time:
I use `rawDefine` with all the constants defined explicitly.

I also found a workaround to get `rawDefine` feature working for components templates by prefixing keys by `_ctx.`
Now i have a problem with tree shaking of components that works, but only for a component tag (not component itself).

**Comment by freddy38510** [comment]
So, the `v-if` directive should have the value `false` instead of `undefined` for complete tree shaking.

What happens if you wrap your component in a `<template>` tag containing the `v-if` directive ?

Or you could try to wrap your component inside a [render function component](https://vuejs.org/guide/extras/render-function.html) or a [functional component](https://vuejs.org/guide/extras/render-function.html#functional-components):

```js
import { h } from 'vue'
import Foo from './Foo.vue'

export default {
  setup() {
    return () => h('template', [SOME_FEATURE ? h(Foo) : undefined])
  }
}
```

**Reply by ohchi** [reply]
> So, the `v-if` directive should have the value `false` instead of `undefined` for complete tree shaking.

My `v-if` directives do have value `false`.

> What happens if you wrap your component in a `<template>` tag containing the `v-if` directive ?

I don't know exactly, but i printed the input of `vite`'s define plugin and found out that `SOME_FEATURE` from `v-if="SOME_FEATURE"` becomes `_ctx.SOME_FEATURE`.
So, with:
```
rawDefine: {
  SOME_FEATURE: false,
  '_ctx.SOME_FEATURE': false
}
```
tree shaking works, but not completely. Tag containing `v-if` is stripped, the component that the tag is referencing isn't.

**Comment by freddy38510** [comment]
Would you have a minimal repro with your latest breakthroughs ?

I'm not quite sure what the problem is. Is it the fact that the component, although removed from its parent component, is still compiled and present in the bundle?

I've just tried an other approach and it seems to work, the Foo component is completely treeshaked:

```vue
<template>
  <Foo />
</template>

<script setup lang="ts">
import { defineAsyncComponent } from 'vue';

const Foo = FEATURE
  ? defineAsyncComponent(() => import('./Foo.vue'))
  : () => '<template>';
</script>
```

```js
// quasar.config.js
// skipping code for clarity

rawDefine: {
  FEATURE: false,
},
```

**Reply by ohchi** [reply]
> Would you have a minimal repro with your latest breakthroughs ?

Yes, i'll do it.

> Is it the fact that the component, although removed from its parent component, is still compiled and present in the bundle?

Yes, it is.

> I've just tried an other approach and it seems to work, the Foo component is completely treeshaked

I think this approach definitely works and can be considered as a fallback option if i cannot manage with a static component.

**Reply by freddy38510** [reply]
I've just found a workaround by using the `/* @__PURE__ */` annotation. It works unless the component you want to be treeshaked imports another component that doesn't have the `/* @__PURE__ */` annotation.

```vue
<template>
  <Foo v-if="false" />
</template>

<script setup lang="ts">
import Foo from './Foo.vue';
</script>
```

```vue
<template>
  <div></div>
</template>

<script lang="ts">
import { defineComponent } from 'vue';

const Foo = /* @__PURE__ */ defineComponent({
  name: 'FooComponent',
});

export default Foo;
</script>
```

It's far from perfect, but at least it's something [to dig into](https://rollupjs.org/configuration-options/#treeshake-annotations).

**Reply by ohchi** [reply]
> I've just found a workaround by using the /* @__PURE__ */ annotation.

`v-if="false"` works for me without dynamic components and `/* @__PURE__ */`
See the problem details in `./src/layouts/MainLayout.vue` of my [repro](https://stackblitz.com/edit/quasarframework-yj7zow?file=README.md)

**Reply by freddy38510** [reply]
With `v-if="false"`, after running the `quasar build` command, I can see the EssentialLink component bundled in `dist/spa/assets/MainLayout.[hash].js` file.

**Reply by ohchi** [reply]
Yes, you're right. Strange, I checked more than once and was completely sure...
So, then the solution i like is with `defineAsyncComponent` that you proposed.
The question is what do you think? Is the workaround with `_ctx.` in `rawDefine` reliable enough?

**Reply by freddy38510** [reply]
This is the simplest solution for proper treeshaking without too much effort.

The only drawback I can see is that the client browser will potentially have to load a large number of components. If one of these component takes a long time to load, you'll have to display a [Loading component](https://vuejs.org/guide/components/async.html#loading-and-error-states). This could degrade the user experience.

About `_ctx.`, at first sight it comes from the [`@vue/compiler-core` package](https://github.com/vuejs/core/blob/main/packages/compiler-core/src/options.ts#L138). I don't know if it's wise to rely on this.

But since you'll no longer use your global constants in SFC template (thanks to async components), you can remove this prefix.

**Reply by ohchi** [reply]
> But since you'll no longer use your global constants in SFC template (thanks to async components), you can remove this prefix.

I'm not gonna give up using constants in SFC template because `v-if` may be used for treeshaking html layout inside a component, not only for tree shaking of child components. I plan to use both these approaches. Dynamic component approach won't be used for most common components for simplicity, but for some kind of specific ones it will.

**Reply by freddy38510** [reply]
In this case, as far as reliability is concerned, it will work until the Vue compiler changes/deletes this prefix.

If you're worried about this, you can manually replace your global constants using your own vite plugin that runs before the `@vitejs/plugin-vue` plugin. You'll receive the uncompiled component as input in the transform hook.

As a minimal proof of concept:
```js
// quasar.config.js

build: {
  extendViteConf (viteConf) {
    viteConf.plugins.push({
      name: "plugin-name",
      enforce: "pre", // make it run before other plugins
      transform(code) {
        return code.replace('SOME_FEATURE', 'false');
     }
    })
  }
}
```

In fact, replacing your global constants before your Vue SFC components are compiled might solves your treeshaking issues.

**Reply by ohchi** [reply, marked as answer]
Thank you for the answers!

The summary is:

1. I use `rawDefine` property of `quasar.config.js` instead of `env`.
2. For every constant in `rawDefine` I make its prefixed with `_ctx.` version (at least it works for current version of vue compiler).
3. And for treeshaking of components I use dynamic components as @freddy38510  proposed [here](https://github.com/quasarframework/quasar/discussions/15998#discussioncomment-6365730)
